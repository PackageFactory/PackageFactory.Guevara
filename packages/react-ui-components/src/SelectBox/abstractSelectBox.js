import React, {PureComponent, PropTypes} from 'react';
import isFunction from 'lodash.isfunction';

export const propTypes = {
    /**
     * This prop represents either a set of options or a function that returns those.
     * Each option must have a value and can have a label and an icon.
     */
    options: React.PropTypes.oneOfType([
        PropTypes.arrayOf(
            PropTypes.shape({
                icon: PropTypes.string,
                value: PropTypes.oneOfType([
                    PropTypes.string,
                    PropTypes.object
                ]).isRequired,
                label: PropTypes.oneOfType([
                    PropTypes.string,
                    PropTypes.object
                ]).isRequired
            })
        ),
        PropTypes.func
    ]),

    /**
     * This prop represents if the options should be loaded on init or when the user starts typing
     * something in the search field.
     */
    loadOptionsOnInput: PropTypes.bool,

    /**
     * This prop is the placeholder text which is displayed in the selectbox when no option was selected.
     */
    placeholder: PropTypes.string,

    /**
     * This prop is an icon for the placeholder.
     */
    placeholderIcon: PropTypes.string,

    /**
     * This prop gets called when an option was selected. It returns the new value.
     */
    onSelect: PropTypes.func.isRequired,

    /**
     * If passed, a `delete` icon will be rendered instead of a chevron,
     * this prop will be called when clicking on the icon.
     */
    onDelete: PropTypes.func,

    /**
     * The minimum amount of items in the select before showing a search box, if set to -1 the search box
     * will never be shown.
     */
    minimumResultsForSearch: PropTypes.number
};

export const state = {
    value: undefined,
    options: undefined,
    isLoadingOptions: false,
};

export default class AbstractSelectBox extends PureComponent {
    static propTypes = {
        ...propTypes
    };

    constructor(...args) {
        super(...args);

        this.handleOptionsLoad = this.handleOptionsLoad.bind(this);
        this.setLoadedOptions = this.setLoadedOptions.bind(this);
        this.select = this.select.bind(this);
    }

    /**
     * returns the options
     *
     * @returns {Array}
     */
    getOptions() {
        return this.state.options || this.props.options;
    }

    /**
     * Loads async options if options is a function.
     *
     * @return {boolean} loadedOptions  TRUE, if options was called (async), otherwise FALSE
     */
    loadOptions() {
        const options = this.props.options;

        this.setState({
            isLoadingOptions: true
        });

        // We only use the current value when loading options, if the
        // 'loadOptionsOnInput' flag is on.
        return isFunction(options) && options({
            value: this.props.loadOptionsOnInput ? this.props.value : undefined,
            callback: this.handleOptionsLoad
        });
    }

    /**
     * Handler for the async options loading callback
     * @param {object} options
     */
    handleOptionsLoad(options) {
        this.setLoadedOptions(options);

        // After options loaded, re-try to select the current element!
        this.select(this.props.value, false);
    }

    /**
     * Sets a given set of options to the internal state.
     * Also disables the loading state.
     *
     * @param options
     */
    setLoadedOptions(options) {
        this.setState({
            options: options,
            isLoadingOptions: false
        });
    }

    select(incomingValue, shouldTriggerOnSelect = true) {
        console.error ('This method needs to be overridden!')
    }

    getOptionLabelForValue(value) {
        return Object.prototype.toString.call(this.getOptions()) === '[object Array]' ?
            this.getOptions().filter(o => o.value === value).map(o => o.label)[0] : null;
    }

    getOptionIconForValue(value) {
        return Object.prototype.toString.call(this.getOptions()) === '[object Array]' ?
            this.getOptions().filter(o => o.value === value).map(o => o.icon)[0] : null;
    }

}
