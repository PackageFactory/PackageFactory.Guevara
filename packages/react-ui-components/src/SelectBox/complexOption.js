
import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import SelectBoxOption from './selectBoxOption';
import style from './style.css';

export default class ComplexOption extends PureComponent {
    static propTypes = {
        option: PropTypes.object.isRequired,
        label: PropTypes.string.isRequired,
        imageUri: PropTypes.string,
        secondaryLabel: PropTypes.string,
        tertiaryLabel: PropTypes.string,
        icon: PropTypes.string
    };

    render() {
        const image = this.props.imageUri ? <img src={this.props.imageUri} alt={this.props.label} className={style.complexOption__image}/> : '';
        const secondaryLabel = this.props.secondaryLabel ?
            <span className={style.complexOption__secondaryLabel}>{this.props.secondaryLabel}</span> :
            '';
        const tertiaryLabel = this.props.tertiaryLabel ?
            <span className={style.complexOption__tertiaryLabel}>{this.props.tertiaryLabel}</span> :
            '';

        return (
            <SelectBoxOption {...this.props} icon={this.props.icon} className={style.complexOption__item}>
                {image}
                <span>{this.props.label}</span>
                {secondaryLabel}
                {tertiaryLabel}
            </SelectBoxOption>
        );
    }
}
