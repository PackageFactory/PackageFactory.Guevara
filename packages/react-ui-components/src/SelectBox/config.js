import {Keys} from 'react-keydown';

const {ENTER, UP, DOWN} = Keys;

export const keys = [ENTER, UP, DOWN];
