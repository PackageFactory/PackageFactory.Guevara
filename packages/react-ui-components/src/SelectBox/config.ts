import {Keys} from 'react-keydown';

const {ENTER, UP, DOWN, ESC} = Keys;

export const keys: ReadonlyArray<number> = [ENTER, UP, DOWN, ESC];
