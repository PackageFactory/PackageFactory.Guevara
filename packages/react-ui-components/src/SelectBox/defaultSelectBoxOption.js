import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import SelectBoxOption from './selectBoxOption';

export default class DefaultSelectBoxOption extends PureComponent {
    static propTypes = {
        option: PropTypes.shape({
            label: PropTypes.string,
            icon: PropTypes.string
        })
    };

    render() {
        const {option} = this.props;

        return (
            <SelectBoxOption {...this.props} icon={option.icon}>
                <span>{option.label}</span>
            </SelectBoxOption>
        );
    }
}
