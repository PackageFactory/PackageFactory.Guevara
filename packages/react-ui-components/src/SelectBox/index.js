import {themr} from 'react-css-themr';
import identifiers from './../identifiers.js';
import style from './style.css';
import SelectBox from './selectBox.js';

const ThemedSelectBox = themr(identifiers.selectBox, style)(SelectBox);

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import DropDown from './../DropDown/index';
import Icon from './../Icon/index';
import IconButton from './../IconButton/index';
import TextInput from './../TextInput/index';
import SearchableSelectBox from './searchableSelectBox';
import SimpleSelectBox from './simpleSelectBox';

export default injectProps({
    DropDownComponent: DropDown,
    IconComponent: Icon,
    IconButtonComponent: IconButton,
    InputComponent: TextInput,
    SearchableSelectBoxComponent: SearchableSelectBox,
    SimpleSelectBoxComponent: SimpleSelectBox
})(ThemedSelectBox);
