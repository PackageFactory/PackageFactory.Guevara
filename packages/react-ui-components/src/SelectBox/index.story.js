import React from 'react';
import {storiesOf, action} from '@kadira/storybook';
import {withKnobs, text} from '@kadira/storybook-addon-knobs';
import {StoryWrapper} from './../_lib/storyUtils.js';
import SelectBox from './index.js';
import DropDown from './../DropDown/index.js';

const options = [
    {value: 'opt1', label: 'Option 1'},
    {value: 'opt2', label: 'Option 2'},
    {value: 'opt3', label: 'Option 3'}
];

const loadOptions = ({callback, value, searchTerm}) => { // TODO change api -> value / searchTerm
    if (value) {
        // simple search for async options
        const filteredOptions = options.filter((option) => {
            return option.value == value;
        });

        setTimeout(() => (callback(filteredOptions)), 1000);
    } else if (searchTerm) {
        // simple search for async options
        const filteredOptions = options.filter((option) => {
            return option.label.toLowerCase().indexOf(searchTerm.toLowerCase()) !== -1;
        });

        setTimeout(() => (callback(filteredOptions)), 1000);
    } else {
        setTimeout(() => (callback(options)), 1000);
    }
};

storiesOf('SelectBox', module)
    .addDecorator(withKnobs)
    .addWithInfo(
        'default',
        () => (
            <StoryWrapper>
                <SelectBox
                    value={"opt1"}
                    options={options}
                    placeholder={text('Placeholder', 'Select')}
                    placeholderIcon={text('Placeholder icon', 'bookmark')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo( // TODO: fix story, option 2 is selected, dropdown contains only this option
        'async',
        () => (
            <StoryWrapper title="SelectBox">
                <SelectBox
                    value={"opt2"}
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Select')}
                    onSelect={action('onSelect')}
                    />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'inside dropdown',
        () => (
            <StoryWrapper title="SelectBox">
                <DropDown isOpen={true}>
                    <DropDown.Header>
                        Dropdown header
                    </DropDown.Header>
                    <DropDown.Contents>
                        <SelectBox
                            options={options}
                            placeholder={text('Placeholder', 'Select')}
                            placeholderIcon={text('Placeholder icon', 'search')}
                            onSelect={action('onSelect')}
                            />
                    </DropDown.Contents>
                </DropDown>
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'searchable',
        () => (
            <StoryWrapper title="SelectBox">
                <SelectBox
                    options={options}
                    placeholder={text('Placeholder', 'Type to search')}
                    onSelect={action('onSelect')}
                    onDelete={action('onDelete')}
                />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'searchableAsync',
        () => (
            <StoryWrapper title="SelectBox">
                <SelectBox
                    value={"opt1"}
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Type to search')}
                    onSelect={action('onSelect')}
                    onDelete={action('onDelete')}
                />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'searchableAsyncWithLoadOnInputAndValu',
        () => (
            <StoryWrapper title="SelectBox">
                <SelectBox
                    value={"opt1"}
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Type to search')}
                    onSelect={action('onSelect')}
                    onDelete={action('onDelete')}
                    loadOptionsOnInput={true}
                />
            </StoryWrapper>
        ),
        {inline: true}
    )
    .addWithInfo(
        'searchableAsyncWithLoadOnInput',
        () => (
            <StoryWrapper title="SelectBox">
                <SelectBox
                    options={loadOptions}
                    placeholder={text('Placeholder', 'Type to search')}
                    onSelect={action('onSelect')}
                    onDelete={action('onDelete')}
                    loadOptionsOnInput={true}
                />
            </StoryWrapper>
        ),
        {inline: true}
    );
