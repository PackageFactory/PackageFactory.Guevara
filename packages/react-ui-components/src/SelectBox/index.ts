import {themr} from '@friendsofreactjs/react-css-themr';
import keydown from 'react-keydown';

import {keys} from './config';
import identifiers from '../identifiers';
import SelectBox from './selectBox';

import style from './style.css';

const ThemedSelectBox = themr(identifiers.selectBox, style)(SelectBox);
const WithKeys = keydown(keys)(ThemedSelectBox);

export default WithKeys;
