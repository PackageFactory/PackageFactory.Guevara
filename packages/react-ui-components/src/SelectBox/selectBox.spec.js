import sinon from 'sinon';
import {createShallowRenderer, createStubComponent} from './../_lib/testUtils.js';
import SelectBox from './selectBox.js';

const DropDownComponent = createStubComponent();
DropDownComponent.Header = createStubComponent();
DropDownComponent.Contents = createStubComponent();

const defaultProps = {
    options: [],
    onSelect: () => null,
    theme: {},
    DropDownComponent,
    IconComponent: createStubComponent(),
    InputComponent: createStubComponent()
};

const shallow = createShallowRenderer(SelectBox, defaultProps);

test('should render passed options straight away', () => {
    const props = {
        options: [{value: 'foo', label: 'bar'}]
    };

    const selectBox = shallow(props);

    expect(selectBox.instance().loadOptions()).toBeFalsy();
    expect(selectBox.instance().getOptions()).toBe(props.options);
});

test('should have the selected value', () => {
    const props = {
        options: [{value: 'foo', label: 'bar'}]
    };
    const selectBox = shallow(props);

    selectBox.instance().select(props.options[0].value);

    expect(selectBox.instance().state.value).toBe(props.options[0].value);
});

test('should call "onSelect" when selecting a value', () => {
    const props = {
        options: [{value: 'foo', label: 'bar'}],
        onSelect: sinon.spy()
    };
    const selectBox = shallow(props);

    selectBox.instance().select(props.options[0].value);

    expect(props.onSelect.calledOnce).toBeTruthy();
});

test('should not call "onSelect" on initial render when having a value set', () => {
    const props = {
        options: [{value: 'foo', label: 'bar'}],
        value: 'some value',
        onSelect: sinon.spy()
    };
    const selectBox = shallow(props);

    selectBox.instance().componentDidMount();

    expect(props.onSelect.notCalled).toBeTruthy();
});
