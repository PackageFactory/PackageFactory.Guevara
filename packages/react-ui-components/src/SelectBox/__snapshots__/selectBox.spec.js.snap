// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBox/> should render correctly. 1`] = `undefined`;
